fs/dafsserver/dafsserver.cpp

```diff
@@ -1630,15 +1620,90 @@ class CRemoteMarkupReadActivity : public CRemoteStreamReadBaseActivity, implemen
     Owned<const IDynamicFieldValueFetcher> fieldFetcher;
     Owned<const IDynamicTransform> translator;
 
+    // JCSMORE - it would be good if these were cached/reused (can I assume anything using fetcher is single threaded?)
     class CFieldFetcher : public CSimpleInterfaceOf<IDynamicFieldValueFetcher>
     {
         unsigned numInputFields;
         const RtlRecord &recInfo;
         Linked<IColumnProvider> currentMatch;
+        const char **compoundXPaths = nullptr;
+
+        const char *queryCompoundXPath(unsigned fieldNum) const
+        {
+            if (compoundXPaths && compoundXPaths[fieldNum])
+                return compoundXPaths[fieldNum];
+            else
+                return recInfo.queryXPath(fieldNum);
+        }
     public:
         CFieldFetcher(const RtlRecord &_recInfo, IColumnProvider *_currentMatch) : recInfo(_recInfo), currentMatch(_currentMatch)
         {
             numInputFields = recInfo.getNumFields();
+
+            // JCSMORE - should this be done (optionally) when RtlRecord is created?
+            for (unsigned fieldNum=0; fieldNum<numInputFields; fieldNum++)
+            {
+                if (recInfo.queryType(fieldNum)->queryChildType())
+                {
+                    const char *xpath = recInfo.queryXPath(fieldNum);
+                    dbgassertex(xpath);
+                    const char *ptr = xpath;
+                    char *expandedXPathPtr = nullptr;
+                    while (true)
+                    {
+                        if (*ptr == xpathCompoundSeparatorChar)
+                        {
+                            if (!compoundXPaths)
+                            {
+                                compoundXPaths = new const char *[numInputFields];
+                                memset(compoundXPaths, 0, sizeof(const char *)*numInputFields);
+                            }
+
+                            char *&expandedXPath = (char *&)compoundXPaths[fieldNum];
```

Comment:
Slightly unusual coding style.  Why not allocate, and then assign expandedXPath into compoundPaths when finished?

Reply:
ok, will change.